File: java/437.path-sum-iii.java
/*
 * @lc app=leetcode id=437 lang=java
 *
 * [437] Path Sum III
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {

    int total = 0;

    public int pathSum(TreeNode root, int sum) {
        if (root != null) count(root, sum, false);
        return total;
    }

    private void count(TreeNode root, int sum, boolean started) {
        if (root == null) return;
        if (root.val == sum) total++;
        if (!started) {
            count(root.left, sum, false);
            count(root.right, sum, false);
        }
        count(root.left, sum - root.val, true);
        count(root.right, sum - root.val, true);
    }

}

File: go/437.path-sum-iii.go
/*
 * @lc app=leetcode id=437 lang=golang
 *
 * [437] Path Sum III
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func pathSum(root *TreeNode, sum int) int {
	return pathSumRec(root, 0, sum, false)
}

func pathSumRec(root *TreeNode, acc, sum int, started bool) int {
	cnt := 0
    if root == nil {
		return cnt
	}
	acc += root.Val
	if acc == sum {
		cnt += 1
	}
	cnt += pathSumRec(root.Left, acc, sum, true) + pathSumRec(root.Right, acc, sum, true)
	if !started {
		cnt += pathSumRec(root.Left, 0, sum, false) + pathSumRec(root.Right, 0, sum, false)
	}
	return cnt
}
// @lc code=end

